### DISCUSSION TITLE
UUID type for mongo id

### DISCUSSION THREAD

**Discussion by waynehaffenden** [typegoose/typegoose#661] [Q&A]
I'm looking at using UUID v4 as the IDs for all the collections/models in Mongo, any ideas on how I do this?

**Comment by hasezoey** [comment]
next time, please directly use github discussions

---

if you mean `v4` from [`uuid`](https://www.npmjs.com/package/uuid), then you can easily do this by setting your type to `String`:

```ts
class Test {
  @prop({ type: () => String, required: true })
  public _id!: string;
}
// Because you are re-defining the "_id" property, you cannot use ObjectId's anymore (or anything that requires it), also, as long as option "default" is not set, you have to provide the "_id" property every time you create a document
```

**Reply by smolinari** [reply]
Also, if you weren't already aware, if you generate your own IDs, then you will be responsible for setting them everywhere they need to be set (i.e. creation/ saving new docs). Mongoose takes care of this for you, if you use the standard ObjectIds. 

https://mongoosejs.com/docs/guide.html#_id

Scott

**Reply by waynehaffenden** [reply]
@hasezoey Thank you, I'll do this in future :)

@smolinari Is there no way of doing this automatically, like the default property on mongoose schema object but in typegoose?

**Reply by hasezoey** [reply]
> @smolinari Is there no way of doing this automatically, like the default property on mongoose schema object but in typegoose?

actually what @smolinari wrote, i had already in the code example as a comment (look above)
TL;DR: unless you set PropOption [`default`](https://typegoose.github.io/typegoose/docs/api/decorators/prop#default), you have to always provide `_id`

also, i would recommend you use [`validate`](https://typegoose.github.io/typegoose/docs/api/decorators/prop#validate) to validate that your id's are only uuid (not some random string)

**Reply by smolinari** [reply]
Haha.. I didn't read that comment. 😊 

Scott
